lua/ulx/modules/sh/cfc_fps.lua

```diff
@@ -0,0 +1,202 @@
+end
+
+if CLIENT then
+    local color_white, color_red, color_green = Color( 255, 255, 255 ), Color( 255, 0, 0 ), Color( 0, 255, 0 )
+    local header = "------------------------------------------------------------"
+
+    local function colorForFPS( fps )
+        local color = color_white
+        if fps < 30 then
+            color = color_red
+
+        elseif fps > 75 then
+            color = color_green
+
+        end
+        return color
+    end
+
+    net.Receive( "CFC_ULX_FPSCheck_ConsoleResults", function()
+        local results = net.ReadTable()
+        MsgC( color_white, header, "\n" )
+        for ply, data in pairs( results ) do
+            local plyColor = team.GetColor( ply:Team() )
+            MsgC( plyColor, ply:Name(), color_white, "\nAverage FPS: ", colorForFPS( data.average ), data.average, color_white, "  Max: ", colorForFPS( data.best ), data.best, color_white, "  Worst: ", colorForFPS( data.worst ), data.worst, "\n" )
+        end
+        MsgC( color_white, header, "\n" )
+    end )
+
+    local function processAndSendBackFPS( allTheFPS )
+        if table.Count( allTheFPS ) <= 0 then return end -- :(
+
+        local count = 0
+        local averageFPS = 0
+        local bestFPS = 0
+        local worstFPS = math.huge
+
+        for _, currentFps in ipairs( allTheFPS ) do
+            count = count + 1
+            averageFPS = averageFPS + currentFps
+            bestFPS = currentFps > bestFPS and currentFps or bestFPS
+            worstFPS = currentFps < worstFPS and currentFps or worstFPS
+        end
+
+        averageFPS = averageFPS / count
+
+        averageFPS = math.Round( averageFPS, 2 )
+        bestFPS = math.Round( bestFPS, 2 )
+        worstFPS = math.Round( worstFPS, 2 )
+
+        net.Start( "CFC_ULX_FPSCheck" )
+            net.WriteInt( averageFPS, 16 )
+            net.WriteInt( bestFPS, 16 )
+            net.WriteInt( worstFPS, 16 )
+        net.SendToServer()
+
+        -- debug print, also tells power users what's happening
+        print( "!fps; sent tracked fps. Average " .. averageFPS .. ". Best " .. bestFPS .. ". Worst " .. worstFPS )
+
+        allTheFPS = nil
+    end
+
+    local timerName = "cfc_ulx_fps_tracker"
+    local function trackFpsFor( trackTime )
+        print( "!fps; tracking fps for " .. trackTime .. " seconds" )
+
+        local allDoneTime = CurTime() + trackTime
+        local allTheFPS = {}
+        local gotData
+
+        timer.Create( timerName, 0.1, 0, function()
+            if allDoneTime < CurTime() then
+                if gotData then
+                    processAndSendBackFPS( allTheFPS )
+                else
+                    print( "!fps; got no fps data, client was tabbed out" )
+                end
+
+                timer.Remove( timerName )
+                return
+            end
+
+            if not system.HasFocus() then return end -- junk data
+
+            local fps = 1 / RealFrameTime()
+            table.insert( allTheFPS, fps )
```

Comment:
Instead of storing a table, you should have 4 local vars to track the min, max, sum, and sample count. That way you can just get `avg = sum / count` when sending back to the server. Storing all those samples in a table will create a bunch of garbage memory.

Reply:
applied